charmcraft/parts.py

```diff
@@ -41,6 +42,57 @@ class CharmPluginProperties(plugins.PluginProperties, plugins.PluginModel):
     charm_python_packages: List[str] = []
     charm_requirements: List[str] = []
 
+    @pydantic.validator("charm_entrypoint", pre=True)
+    def validate_entry_point(cls, charm_entrypoint, values):
+        """Validate the entry point."""
+        # the location of the project is needed
+        if "source" not in values:
+            raise ValueError(
+                "cannot validate 'charm-entrypoint' because invalid 'source' configuration"
+            )
+        project_dirpath = pathlib.Path(values["source"]).resolve()
+
+        # check that the entrypoint, is inside the project
+        filepath = (project_dirpath / charm_entrypoint).resolve()
+        if project_dirpath not in filepath.parents:
+            raise ValueError(
+                "charm entry point must be inside the project: {!r}".format(str(filepath))
+            )
+
+        # store the entrypoint always relative to the project's path (no matter if the origin
+        # was relative or absolute)
+        rel_entrypoint = (project_dirpath / charm_entrypoint).relative_to(project_dirpath)
+        charm_entrypoint = rel_entrypoint.as_posix()
+
+        return charm_entrypoint
+
+    @pydantic.validator("charm_requirements", pre=True, always=True)
+    def validate_requirements(cls, charm_requirements, values):
+        """Validate the specified requirement or dynamically default it.
+
+        The default is dynamic because it's only requirements.txt if the
+        file is there.
+        """
+        # the location of the project is needed
+        if "source" not in values:
```

Comment:
Isn't `source` guaranteed to be declared at this point, since it's added during parts pre-validation?

Reply:
No, because the "implicit source" is only triggered for the case of "standard charm plugin usage".

IOW, if you don't declare `parts`, or do not declare a charm part, or use `parts` with a `charm` plugin, etc, the system will complete the `source` if not explicitly indicated (pointing to the project in disk). However, if you use the `charm` plugin in a different part and do not declare a `source`, it will not be completed...